crates/storage/db/src/snapshot.rs

```diff
@@ -0,0 +1,105 @@
+//! reth's snapshot creation from database tables
+
+use crate::{
+    abstraction::cursor::DbCursorRO,
+    table::{Key, Table},
+    transaction::DbTx,
+    DatabaseError, RawKey, RawTable,
+};
+use reth_interfaces::{RethResult};
+use reth_nippy_jar::NippyJar;
+use std::ops::{RangeBounds, RangeInclusive};
+
+/// Macro that generates snapshot creation functions that take an arbitratry number of [`Table`] and
+/// creates a [`NippyJar`] file out of their [`Table::Value`]. Each list of [`Table::Value`] from a
+/// table is a column of values.
+///
+/// Has membership filter set and compression dictionary support.
+macro_rules! generate_snapshot_func {
+    ($(($($tbl:ident),+)),+ $(,)? ) => {
+        $(
+            paste::item! {
+                /// Creates a snapshot from specified tables. Each table's `Value` iterator represents a column.
+                ///
+                /// **Ensure the range contains the same number of rows.**
+                ///
+                /// * `tx`: Database transaction.
+                /// * `range`: Data range for columns in tables.
+                /// * `dict_compression_set`: Sets of column data for compression dictionaries. Max size is 2GB. Row count is independent.
+                /// * `use_hash`: Flag to indicate whether to create snapshot filters/phf from `TxHash` or `BlockHash`
+                /// * `row_count`: Total rows to add to `NippyJar`. Must match row count in `range`.
+                /// * `nippy_jar`: Snapshot object responsible for file generation.
+                #[allow(non_snake_case)]
+                pub fn [<create_snapshot$(_ $tbl)+>]<'tx,
+                    $($tbl: Table<Key=K>,)+
+                    HashTbl: Table<Key=K>,
+                    K
+                >
+                (
+                    tx: &impl DbTx<'tx>,
+                    range: RangeInclusive<K>,
+                    dict_compression_set: Option<Vec<impl Iterator<Item = Vec<u8>>>>,
+                    use_hash: bool,
+                    row_count: usize,
+                    nippy_jar: &mut NippyJar
+                ) -> RethResult<()>
+                    where K: Key + Copy
+                {
+                    let range: RangeInclusive<RawKey<K>> = RawKey::new(*range.start())..=RawKey::new(*range.end());
+
+                    handle_compression_and_indexing::<HashTbl>(
+                        tx, dict_compression_set, use_hash.then_some(range.clone()), nippy_jar
+                    )?;
+
+                    // Creates the cursors for the columns
+                    $(
+                        let mut [< $tbl _cursor>] = tx.cursor_read::<RawTable<$tbl>>()?;
+                        let [< $tbl _iter>] = [< $tbl _cursor>]
+                            .walk_range(range.clone())?
+                            .into_iter()
+                            .map(|row| row.map(|(_key, val)| val.take()).expect("exist in database"));
```

Comment:
I'm not sure how to handle this. Reminder that we want this to remain an iterator, but I'd rather not change the `NippyJar::freeze` to accommodate a `Item = Result<Vec<u8>>`... thoughts?